### ISSUE TITLE
[22.0] `stellar keys generate` should no longer fund

### ISSUE THREAD

**Issue by chadoh** [stellar/stellar-cli#1407]
(Original reported issue by @chadoh below)

Currently, `stellar keys generate` not only generate keys, but also funds them. This is a bit confusing for the user (e.g. the original error Chad has encountered), and makes thing more complicated because same thing can be done in multiple ways.
The proposal is to remove fund functionality from `keys generate`, as well as all network related flags (no network connection is needed anymore, we just generating key pair), and `--no-fund` (this is a default behavior now).
User can still fund keys with `stellar keys fund`

Because this is a breaking change, we can't make it before 22.0

 ---

### What version are you using?

```
$ stellar --version
stellar 21.0.0 (c558dd92613d91a34011cf5105869a5f57af230c)
soroban-env 21.1.0 (8d76e4037417b80dc25c979cc44b8e1281e8ad84)
soroban-env interface version 90194313216
stellar-xdr 21.1.0 (4dcbb918edcd793fdf064b3ee28ceacc6bd9d1bb)
xdr curr (70180d5e8d9caee9e8645ed8a38c36a8cf403cd9)
```

### What did you do?

    stellar keys generate alice

<img width="1297" alt="image" src="https://github.com/stellar/stellar-cli/assets/221614/eaf1c511-4dfc-4cd2-b18b-e60965344be9">

### What did you expect to see?

If no network provided, should generate keys and quietly skip funding.

If I provide `--no-fund`, same behavior.

### What did you see instead?

Errors unless network provided

**Comment by leighmcculloch** [comment]
Another option is we remove funding from generating. I find it odd that every time I generate a key, I get an account on testnet, when sometimes I don't want an account created either now or not yet.

**Comment by willemneal** [comment]
Yeah perhaps we should flip the flag. This way users who want to have it work in one step can still use just one command.

**Comment by leighmcculloch** [comment]
> If no network provided, should generate keys and quietly skip funding.

A downside of this is that when someone has a network set as an env var, it'd still be picked up there, so I could feel like I'm typing a simple generate command but then get hit with funding because I set the env var for commands that I expected to use the network.

**Comment by Ifropc** [comment]
I think the idea was to flip the logic to always **not** fund as a breaking change in 22.0 
User can still provide `--fund` to fund the new accounts explicitly. WDIT @leighmcculloch ?

**Comment by leighmcculloch** [comment]
Adding a `--fund` flag isn't too different to folks using the `fund` command, so if we're removing the default fund behavior, I'd suggest we leave off the `--fund` flag so there's one way to do the thing rather than two.

**Comment by Ifropc** [comment]
Hmm yeah that makes sense, you can just `stellar keys generate alice && stellar keys fund alice`, which is just a bit more verbose than `stellar keys generate alice --fund`. IMO it makes sense that `generate` only generates the key and doesn't fund it :)

**Comment by willemneal** [comment]
Yeah  we should strive to make each command do one thing as much as possible. I am still a little annoyed that deploy hash --wasm and does the install as well. Being explicit as much as possible helps people understand the steps.

**Comment by fnando** [comment]
I feel the case where keys are funded is more common than the other way around. I believe  we should fix `--no-fund` rather than changing the behavior. 🤔

**Comment by Ifropc** [comment]
I think adding a `--fund` flag is fine. 
I agree that most common use case is to create and then fund the key, but technically speaking key != account and personally it's been confusing to me why we are creating and funding an account (sometimes I just want to create a keypair and that's it)
Adding a flag also goes well with our approach with "command does only one thing, but can also do something extra with a flag -- it equals to piping result to a different command"

**Comment by Ifropc** [comment]
> I feel the case where keys are funded is more common than the other way around. I believe we should fix --no-fund rather than changing the behavior. 🤔

BTW due to how clap flattening works we can't fix it right now, at least not easily 😔  (you can't make group to be optional, which network options is)

**Comment by willemneal** [comment]
One solution is to not have clap handle what is required. Currently there is a `config.get_network` which will throw an error if the args are not provided. 
So we just need to remove `required_unless_present` and then the error will surface only when trying to resolve the network.
https://github.com/stellar/stellar-cli/blob/c8c3871ef3a409328d9b0561f87d55c00bd41071/cmd/soroban-cli/src/config/network.rs#L41-L67

**Comment by willemneal** [comment]
Or in the case where you just have the network args:

https://github.com/stellar/stellar-cli/blob/c8c3871ef3a409328d9b0561f87d55c00bd41071/cmd/soroban-cli/src/config/network.rs#L71-L88

**Comment by Ifropc** [comment]
We can indeed make all network args optional, but then we need to do validation ourselves, which will give a different input compared to clap (don't think it's a major downside, but still)

**Comment by fnando** [comment]
@Ifropc maybe this a good reason for #1616. It's easier to make one single argument optional when needed. 🤔

**Comment by Ifropc** [comment]
^ One of the reasons I want to see that change actually 😆

**Comment by willemneal** [comment]
> We can indeed make all network args optional, but then we need to do validation ourselves, which will give a different input compared to clap (don't think it's a major downside, but still)

I'm saying we are already doing the validation. So removing the clap requires should not break anything, though we might need to improve the error message.
